astropy/cosmology/core.py

```diff
@@ -2110,7 +2053,167 @@ def _analytic_lookback_time(self, z):
+            z = np.asarray(z)
+        Om0, Ode0, Ok0 = self._Om0, self._Ode0, self._Ok0
+        if self._massivenu:
+            Or = self._Ogamma0 * (1 + self.nu_relative_density(z))
+        else:
+            Or = self._Ogamma0 + self._Onu0
+        zp1 = 1.0 + z
+
+        return (zp1 ** 2 * ((Or * zp1 + Om0) * zp1 + Ok0) + Ode0)**(-0.5)
+
+
+class FlatLambdaCDM(LambdaCDM):
+    """FLRW cosmology with a cosmological constant and no curvature.
+
+    This has no additional attributes beyond those of FLRW.
+
+    Parameters
+    ----------
+    H0 : float or `~astropy.units.Quantity`
+        Hubble constant at z = 0.  If a float, must be in [km/sec/Mpc]
+
+    Om0 : float
+        Omega matter: density of non-relativistic matter in units of the
+        critical density at z=0.
+
+    Tcmb0 : float or scalar `~astropy.units.Quantity`, optional
+        Temperature of the CMB z=0. If a float, must be in [K].
+        Default: 0 [K]. Setting this to zero will turn off both photons
+        and neutrinos (even massive ones).
+
+    Neff : float, optional
+        Effective number of Neutrino species. Default 3.04.
+
+    m_nu : `~astropy.units.Quantity`, optional
+        Mass of each neutrino species. If this is a scalar Quantity, then all
+        neutrino species are assumed to have that mass. Otherwise, the mass of
+        each species. The actual number of neutrino species (and hence the
+        number of elements of m_nu if it is not scalar) must be the floor of
+        Neff. Typically this means you should provide three neutrino masses
+        unless you are considering something like a sterile neutrino.
+
+    Ob0 : float or None, optional
+        Omega baryons: density of baryonic matter in units of the critical
+        density at z=0.  If this is set to None (the default), any
+        computation that requires its value will raise an exception.
+
+    name : str, optional
+        Name for this cosmological object.
+
+    Examples
+    --------
+    >>> from astropy.cosmology import FlatLambdaCDM
+    >>> cosmo = FlatLambdaCDM(H0=70, Om0=0.3)
+
+    The comoving distance in Mpc at redshift z:
+
+    >>> z = 0.5
+    >>> dc = cosmo.comoving_distance(z)
+    """
+
+    def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+                 m_nu=u.Quantity(0.0, u.eV), Ob0=None, name=None):
+
+        LambdaCDM.__init__(self, H0, Om0, 0.0, Tcmb0, Neff, m_nu, name=name,
+                           Ob0=Ob0)
+        # Do some twiddling after the fact to get flatness
+        self._Ode0 = 1.0 - self._Om0 - self._Ogamma0 - self._Onu0
+        self._Ok0 = 0.0
+
+        # Please see "Notes about speeding up integrals" for discussion
+        # about what is being done here.
+        if self._Tcmb0.value == 0:
+            self._inv_efunc_scalar = scalar_inv_efuncs.flcdm_inv_efunc_norel
+            self._inv_efunc_scalar_args = (self._Om0, self._Ode0)
+            # Repeat the optimization reassignments here because the init
```

Comment:
Is this really necessary?  The tweaking doesn't look like it modifies any of the variables your repeat uses.

Reply:
`FlatLambdaCDM` calls `LambdaCDM.__init__` with a non-flat Universe:

```
LambdaCDM.__init__(self, H0, Om0, 0.0, Tcmb0, Neff, m_nu, name=name,
                           Ob0=Ob0)
```

It just passes `Ode0=0.0` and then fixes up later.  So the `LambdaCDM.__init__` doesn't set the optimizations appropriate for a flat Universe.

Passing `Ode0=1-Om0` to `LambdaCDM.__init__` wouldn't be right either in case of non-zero radiation density.  While the *current* set of optimizations in `LambdaCDM` are only valid for zero radiation density, in which case passing `Om0=Om0, Ode0=1-Om0` would create a flat Universe, this seems fragile.

Reply:
Ok, I see what you mean.  This still seams... painful.  What would you think about creating a function that
did these special case assignments inside LambdaCDM and then calling it from FlatLambdaCDM?

Reply:
Good suggestion!  Done.  Created `LambdaCDM._optimize_flat_norad` that does the reassignments for the different optimization cases.